webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+				rdb.RPush(ctx, firstTimeQueue, msgBytes)
+				rdb.LTrim(ctx, firstTimeQueue, 0, maxQueueSize-1)
+			} else {
+				// Known bucket, send to bucket queue
+				queueKey := bucketQueuePrefix + bucket
+				rdb.RPush(ctx, queueKey, msgBytes)
+				rdb.LTrim(ctx, queueKey, 0, maxQueueSize-1)
+
+				// Start workers if new bucket
+				if _, exists := GetWorker(bucket, &mu); !exists {
+					worker := NewBucketWorker(bucket, rdb)
+					SetWorker(bucket, worker, &mu)
+					go worker.ProcessQueue(rdb)
+				}
+			}
+		}
+	}()
+
+	go ProcessFirstTime(rdb, &mu)
+	go SubscribeTransients(rdb, &mu)
+
+	// Keep running
+	select {}
+}
+
+// SetupRedis initializes the Redis client
+func SetupRedis() *redis.Client {
+	logger.Info("Initializing redis")
+	host := os.Getenv("HLL_REDIS_HOST")
+	if host == "" {
+		panic("HLL_REDIS_HOST not set")
+	}
+
+	portStr := os.Getenv("HLL_REDIS_PORT")
+	port, err := strconv.Atoi(portStr)
+	if err != nil || port <= 0 {
+		panic("HLL_REDIS_PORT not set or invalid")
+	}
+
+	addr := fmt.Sprintf("%s:%d", host, port)
+	return redis.NewClient(&redis.Options{
+		Addr: addr,
+		DB:   0, // CRCON uses DB #0 as a global shared database for all game servers
+	})
+}
+
+func ExtractWebhookID(url string) (string, error) {
+	matches := webhookIDPattern.FindStringSubmatch(url)
+	if len(matches) < 2 {
+		return "", fmt.Errorf("no webhook ID found in URL")
+	}
+	return matches[1], nil
+}
+
+func SetGloballyRateLimited(rdb *redis.Client, resetTime time.Time, resetAfter time.Duration) {
+	ctx := context.Background()
+	val := strconv.FormatInt(resetTime.Unix(), 10)
+	rdb.Set(ctx, discordGloballyRateLimited, val, resetAfter)
+}
+
+// Add a hash entry anytime a bucket is rate limited so we can count
+// the number of rate limits within the user configured window size
+// and return it from the API
+func UpdateBucketRateLimitCount(rdb *redis.Client, bucket string) {
+	ctx := context.Background()
+	hashName := fmt.Sprintf("%s:%s", bucketRateLimitCountHash, bucket)
+	key := strconv.FormatInt(time.Now().Unix(), 10)
+	rdb.HSet(ctx, hashName, key, true)
+	rdb.HExpire(ctx, hashName, rateLimitWindowSize, key)
+}
+
+// Flag a webhook error in Redis so the status is available to the API
+// these are cleared on a successful request to Discord
+func SetWebhookError(rdb *redis.Client, webhookID string, http_401, http_403, http_404 bool) {
+	ctx := context.Background()
+	key := fmt.Sprintf("%s:%s", webhookErrors, webhookID)
+	payload := map[string]bool{
+		"http_401": http_401,
+		"http_403": http_403,
+		"http_404": http_404,
+	}
+	rdb.HSet(ctx, key, payload)
+}
+
+func ClearWebhookError(rdb *redis.Client, webhookID string) {
+	ctx := context.Background()
+	key := fmt.Sprintf("%s:%s", webhookErrors, webhookID)
+	payload := map[string]bool{
+		"http_401": false,
+		"http_403": false,
+		"http_404": false,
+	}
+	rdb.HSet(ctx, key, payload)
+}
+
+func getGitTag() string {
+	// Run git describe --tags
+	cmd := exec.Command("git", "describe", "--tags")
```

Comment:
tbh, I don't think that the whole git repo should be in the final image. Instead, what about using `git describe --tags` in the build stage of the container and set the tag as an environment variable backed into the image? Then this would be a simple `os.Getenv` call.